walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,95 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Iterator;
 
 public final class FindMeetingQuery {
+
+  /** 
+   * Return all possible time ranges to hold a meeting of specific duration.
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Find which times work for mandatory attendees
+    ArrayList<TimeRange> busyRangesForMandatoryAttendees = findBusyRanges(request.getAttendees(), events);
+    Collections.sort(busyRangesForMandatoryAttendees, TimeRange.ORDER_BY_START);
+    Collection<TimeRange> freeRangesForMandatoryAttendees = findFreeRanges(request, busyRangesForMandatoryAttendees);
+    
+    // Factor in optional attendees if there are any
+    if (!request.getOptionalAttendees().isEmpty()) {
+      ArrayList<TimeRange> busyRangesForOptionalAttendees = findBusyRanges(request.getOptionalAttendees(), events);
+      Collections.sort(busyRangesForOptionalAttendees, TimeRange.ORDER_BY_START);
+      
+      ArrayList<TimeRange> busyRangesForAllAttendees = new ArrayList<>();
+      busyRangesForAllAttendees.addAll(busyRangesForMandatoryAttendees);
+      busyRangesForAllAttendees.addAll(busyRangesForOptionalAttendees);
+      Collections.sort(busyRangesForAllAttendees, TimeRange.ORDER_BY_START);
+      Collection<TimeRange> freeRangesForAllAttendees = findFreeRanges(request, busyRangesForAllAttendees);
+      
+      // Handles case where there are no mandatory attendees but optional attendees have no availability
+      if (freeRangesForAllAttendees.isEmpty() && request.getAttendees().isEmpty()) {
+        return Collections.emptyList();
+      }
+
+
+      // If more than on time slot exists to accommodate both mandatory & optional attendees, return that
+      if (!freeRangesForAllAttendees.isEmpty()) {
+        return freeRangesForAllAttendees;
+      }
+    }
+
+    return freeRangesForMandatoryAttendees;
+  }
+
+  /** 
+   * Returns an ArrayList of all busy TimeRanges.
+   * Check all events for the day and add TimeRanges of events with our mandatory attendees.
+  */
+  private ArrayList<TimeRange> findBusyRanges(Collection<String> attendees, Collection<Event> events) {
```

Comment:
For method return types, you should generally use the immutable type (such as ImmutableList) instead of the general collection interface type (such as List). This communicates to your callers the semantic guarantees of immutable types, which is almost always very useful information. When it's a general collection type, callers may be under the impression that they can mutate the return value, leading to failures at runtime.